### PR TITLE
Use the OS thread name by default if `THREAD_INFO` has not been initialized

### PR DESCRIPTION
**Pull request by ChrisDenton**
Currently if `THREAD_INFO` hasn't been initialized then the name will be set to `None`.  This PR changes it to use the OS thread name by default. This mostly affects foreign threads at the moment but we could expand this to make more use of the OS thread name in the future.

Note: I've only implemented `Thread::get_name` for windows, linux and macos (and macos adjacent) targets. The rest just return `None`.

### CODE REVIEW

File: library/std/src/sys/pal/windows/thread.rs [hunk 1]
```diff
@@ -71,6 +71,28 @@ impl Thread {
...
+                let mut len = 0;
+                while *ptr.add(len) != 0 {
+                    len += 1;
+                }
+                core::slice::from_raw_parts(ptr, len)
+            })
+            .into_bytes();
+            // Attempt to free the memory.
+            // This should never fail but if it does then there's not much we can do about it.
+            debug_assert!(c::LocalFree(ptr.cast::<c_void>()).is_null());
```
Comment (ollie27) [inline library/std/src/sys/pal/windows/thread.rs:91]: Shouldn't this call `LocalFree` even if debug assertions aren't enabled?

Comment (ChrisDenton) [reply library/std/src/sys/pal/windows/thread.rs:91]: Oh, doh! Yeah, I was trying to be too clever by half and really messed that up.

File: library/std/src/sys/pal/unix/thread.rs [hunk 2]
```diff
@@ -225,6 +225,44 @@ impl Thread {
...
+
+    #[cfg(not(any(
+        target_os = "linux",
+        target_os = "macos",
+        target_os = "ios",
+        target_os = "tvos",
+        target_os = "watchos"
+    )))]
+    pub fn get_name() -> Option<CString> {
+        None
```
Comment (cuviper) [inline library/std/src/sys/pal/unix/thread.rs:263]: There are other targets with `libc::pthread_getname_np` too, but I think it's ok to leave them for later.

However, don't we need this kind of placeholder in the other `sys/pal/*/thread.rs` too?

Comment (ChrisDenton) [reply library/std/src/sys/pal/unix/thread.rs:263]: Good point. For some reason I had it in my head that they all used the Unix `Thread` but I don't have a clue why I thought that 😅

File: library/std/src/thread/tests.rs [hunk 3]
```diff
@@ -69,6 +69,25 @@ fn test_named_thread_truncation() {
...
+))]
+#[test]
+fn test_get_os_named_thread() {
+    use crate::sys::thread::Thread;
+    let handler = thread::spawn(|| {
+        let name = c"test me please";
+        Thread::set_name(name);
+        assert_eq!(name, Thread::get_name().unwrap().as_c_str());
+    });
+    handler.join().unwrap();
```
Comment (RalfJung) [inline library/std/src/thread/tests.rs:88]: Is there a particular reason why this test runs in a new thread, rather than the main thread?

Comment (ChrisDenton) [reply library/std/src/thread/tests.rs:88]: I was worried about affecting the thread name for all tests.

Comment (RalfJung) [reply library/std/src/thread/tests.rs:88]: Ah, makes sense. A comment explaining that would have avoided some surprise and confusion here. :)

**Pull request merged**